File: BackEnd/src/domain/interfaces/usecases/chat/ICreateConversationUseCase.ts
/**
 * @file ICreateConversationUseCase.ts
 * @description Interfaz del caso de uso para obtener o crear una conversación entre dos usuarios.
 * Implementa la semántica "get or create": si ya existe la conversación, la devuelve.
 * Registrado en el container con el símbolo TYPES.IGetOrCreateConversationUseCase.
 * @module Chat
 * @layer Domain
 */

import type { ConversationResponseDTO } from '../../../dtos/chat/conversation-response.dto.js';

/**
 * Contrato del use case GetOrCreateConversation.
 * Garantiza que solo exista una conversación entre el par de usuarios.
 */
export interface ICreateConversationUseCase {
  /**
   * Ejecuta el caso de uso.
   *
   * @param participantId — ID del otro usuario con quien iniciar la conversación.
   * @param userId — ID del usuario autenticado.
   * @returns ConversationResponseDTO de la conversación (existente o nueva).
   * @throws {NotFoundException} Si el otro participante no existe.
   */
  execute(participantId: string, userId: string): Promise<ConversationResponseDTO>;
}

File: FrontEnd/lib/domain/interfaces/usecases/chat/i_create_conversation_use_case.dart
/// @file i_create_conversation_use_case.dart
/// @description Interfaz: Crea o recupera una conversación 1:1 con otro usuario.
/// @module Chat
/// @layer Domain
library;

import '../../../entities/conversation.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// I CREATE CONVERSATION USE CASE
// ═══════════════════════════════════════════════════════════════════════════════

/// Contrato del use case que obtiene o crea una conversación 1:1 idempotentemente.
abstract interface class ICreateConversationUseCase {
  /// Devuelve la conversación existente con [otherUserId], o la crea si no existe.
  ///
  /// [param] otherUserId — ID del otro participante.
  /// [returns] [Conversation] existente o recién creada.
  /// [throws] AppError — si la operación falla.
  Future<Conversation> execute(String otherUserId);
}
